### PR TITLE
[NFC] Define InnerSymbol OpInterface and document rational in op

### PR DESCRIPTION
**Pull request by darthscsi**

### CODE REVIEW

**Comment by teqdruid** [comment]
This seems more generally applicable that FIRRTL, right? If so, I think it should go in the hw dialect.

**Comment by darthscsi** [comment]
I was planning on migrating it once it was done if it proved useful.

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h [hunk 1]
```diff
@@ -63,4 +64,14 @@ LogicalResult verifyModuleLikeOpInterface(FModuleLike module);
 } // namespace firrtl
 } // namespace circt
 
+namespace mlir {
```
Comment (teqdruid) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h:67]: Is this not generate-able by tblgen?

**Review by teqdruid — APPROVED** [review]
We should also add `InnerSymbolAttribute` and `InnerSymbolRefAttribute`.

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 2]
```diff
@@ -275,4 +275,36 @@ def FConnectLike : OpInterface<"FConnectLike"> {
...
+    The second problem arrises from nesting.  Symbols defining 
+    ops must be immediate children of a symbol table.  FIRRTL
+    and SV operations which define a inner_sym are grandchildren,
+    at least, of a symbol table and may be much further nested.
+    Lastly, ports need to define inner_sym, something not allowed
+    by normal symbols.
+  }];
+
+  let cppNamespace = "::circt::firrtl";
+    let methods = [];
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:306]: Nit: indentation: 
```suggestion
  let methods = [];
```
Also, should this be defining a method like `inner_sym` or `inner_symAttr`?

**Review by seldridge — APPROVED** [review]

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 3]
```diff
@@ -275,4 +275,36 @@ def FConnectLike : OpInterface<"FConnectLike"> {
...
   ];
-}
\ No newline at end of file
+}
+
+def InnerSymbol : OpInterface<"InnerSymbolOpInterface"> {
+  let description = [{
+    This interface describes an operation that may define an
+    `inner_sym`.  An `inner_sym` operation resides 
+    in arbitrarily-nested regions of a region that define a
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:284]:
```suggestion
    in arbitrarily-nested regions of a region that defines a
```

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 4]
```diff
@@ -275,4 +275,36 @@ def FConnectLike : OpInterface<"FConnectLike"> {
...
+    symbols are resolved by first going up to the closest
+    parent symbol table and resolving from there (recursing
+    down for complex symbol paths).  In FIRRTL and SV, modules
+    define a symbol in a circuit or std.module symbol table.
+    For instances to be able to resolve the modules they
+    instantiate, the symbol use in an instance must resolve 
+    in the top-level symbol table.  If a module were a
+    symbol table, modules would try to resolve from their
+    own module, never seeing other modules.
+    The second problem arrises from nesting.  Symbols defining
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:297]:
```suggestion
    The second problem arises from nesting.  Symbols defining 
```

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 5]
```diff
@@ -275,4 +275,36 @@ def FConnectLike : OpInterface<"FConnectLike"> {
...
+    Inner Symbols are different from normal symbols due to 
+    MLIR symbol table resolution rules.  Specifically normal
+    symbols are resolved by first going up to the closest
+    parent symbol table and resolving from there (recursing
+    down for complex symbol paths).  In FIRRTL and SV, modules
+    define a symbol in a circuit or std.module symbol table.
+    For instances to be able to resolve the modules they
+    instantiate, the symbol use in an instance must resolve 
+    in the top-level symbol table.  If a module were a
+    symbol table, modules would try to resolve from their
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:295]: "If a module were a symbol table, modules would try", maybe something like:
"If a module was a symbol table, symbol references within the module would try"?

**Review by dtzSiFive — COMMENTED** [review]
This is a great step towards better handling of `inner_sym`, thanks!

Added some suggestions for typos/nits :).

**Pull request merged**